streaming/base/util.py

```diff
@@ -196,3 +204,138 @@ def get_import_exception_message(package_name: str, extra_deps: str) -> str:
+                download_timeout: int = 100) -> int:
+    """Merge index.json from a list of remote or local directories.
+
+    Args:
+        folder_urls (Iterable): folders that contain index.json for the partition
+            each element can take the form of a single path string or a tuple string
+
+            for each  url in folder_urls, if url is
+                1. tuple (local, remote): check if local is accessible.
+                    -> Yes: use local index to merge
+                    -> No:  download from remote first, then merge
+                2. str (local path): use local path to merge.
+                    raise FileNotFoundError if any local index is not accessible
+                3. str (remote url): download to a temp directory first, then merge
+
+        out (Union[str, Tuple[str, str]]): path to put the merged index file
+        keep_local (bool): Keep local copy of the merged index file. Defaults to ``True``
+        overwrite (bool): Overwrite merged index file in out if there exists one.Defaults to ``True``
+        download_timeout (int): The allowed time for downloading each json file
+            defaults to 60, same as streaming.download_file
+
+    Returns:
+        int: count of files downloaded during function call
+    """
+    # Import here to avoid circular import error
+    from streaming.base.storage.download import download_file
+    from streaming.base.storage.upload import CloudUploader
+
+    if not folder_urls:
+        logger.warning('No partitions exist, no index merged')
+        return 0
+
+    # This is the index json file name, e.g., it is index.json as of 0.6.0
+    index_basename = get_index_basename()
+
+    cu = CloudUploader.get(out, keep_local=True, exist_ok=True)
+    if os.path.exists(os.path.join(cu.local, index_basename)) and overwrite:
+        logger.warning('Merged index already exists locally. no index merged if overwrite=False')
+        return 0
+
+    # Remove '/' from right, so os.path.basename gives relative path to each folder
+    urls = []
+    for url in folder_urls:
+        if type(url) is str:
+            urls.append(url.rstrip('/').strip())
+        else:
+            urls.append((url[0].rstrip('/').strip(), url[1].rstrip('/').strip()))
+
+    # Determine if we need to call download_file.
+    download = False
+    for url in urls:
+        if type(url) is tuple:
+            # If driver cannot access the local path, download = True
+            download = not os.path.exists(url[0])
+        else:
+            # If url is a remote, download = True, False otherwise
+            download = urllib.parse.urlparse(url).scheme != ''
+
+        # As long as one index file needs download, we download them all to keep it simple
+        if download:
+            break
+
+    print('download = ', download)
+    # Prepare a temp folder to download index.json rom remote if necessary. Removed in the end.
+    with tempfile.TemporaryDirectory() as temp_root:
```

Comment:
Would recommend logging the value of `temp_root` path so that the user aware of the path if passing just remote path.

Reply:
Done.